src/adjancency_matrix.jl

```diff
@@ -0,0 +1,166 @@
+"""
+function AdjacencyMatrix(
+    branches,
+    nodes::Vector{PSY.Bus};
+    check_connectivity::Bool = true,
+    kwargs...,
+)
+    a, bus_lookup = calculate_adjacency(branches, nodes)
+    bus_ax = PSY.get_number.(nodes)
+    axes = (bus_ax, bus_ax)
+    look_up = (bus_lookup, bus_lookup)
+
+    if check_connectivity
+        connected = _validate_connectivity(a, bus_lookup, dfs_connectivity)
+        !connected && throw(DataFormatError("Network not connected"))
+    end
+
+    return AdjacencyMatrix(a, axes, look_up, find_slack_positions(nodes))
+end
+
+function validate_connectivity(
+    M::AdjacencyMatrix;
+    connectivity_method::Function = goderya_connectivity,
+)
+    return _validate_connectivity(M.data, M.lookup[1], connectivity_method)
+end
+
+function _validate_connectivity(
+    M::SparseArrays.SparseMatrixCSC{Int8, Int},
+    bus_lookup::Dict{Int, Int},
+    connectivity_method::Function,
+)
+    connected = connectivity_method(M, bus_lookup)
+    return connected
+end
+
+function _goderya(ybus::SparseArrays.SparseMatrixCSC)
+    node_count = size(ybus)[1]
+    max_I = node_count^2
+    I, J, val = SparseArrays.findnz(ybus)
+    T = SparseArrays.sparse(I, J, ones(Int8, length(val)))
+    T_ = T * T
+    for n in 1:(node_count - 1)
+        I, _, _ = SparseArrays.findnz(T_)
+        if length(I) == max_I
+            @info "The System has no islands"
+            break
+        elseif length(I) < max_I
+            temp = T_ * T
+            I_temp, _, _ = SparseArrays.findnz(temp)
+            if all(I_temp == I)
+                @warn "The system contains islands" maxlog = 1
+            end
+            T_ = temp
+        else
+            @assert false
+        end
+        #@assert n < node_count - 1
+    end
+    return I
+end
+
+function goderya_connectivity(M::SparseArrays.SparseMatrixCSC{Int8, Int},
+    bus_lookup::Dict{Int, Int})
+    @info "Validating connectivity with Goderya algorithm"
+    node_count = length(bus_lookup)
+    node_count > 15_000 &&
+        @warn "The Goderya algorithm is memory intensive on large networks and may not scale well, try `connectivity_method = dfs_connectivity"
+
+    I = _goderya(M)
+
+    connections = Dict([i => count(x -> x == i, I) for i in Set(I)])
+
+    if length(Set(I)) == node_count
+        connected = true
+        if any(values(connections) .!= node_count)
+            cc = Set(values(connections))
+            @warn "Network has at least $(length(cc)) connected components with $cc nodes"
+            connected = false
+        end
+    else
+        connected = false
+    end
+    return connected
+end
+
+function dfs_connectivity(M::SparseArrays.SparseMatrixCSC,
+    bus_lookup::Dict{Int, Int})
+    @info "Validating connectivity with depth first search (network traversal)"
+    cc = find_connected_components(M, bus_lookup)
+    if length(cc) != 1
+        @warn "Network has at least $(length(cc)) connected components with $(length.(cc)) nodes"
```

Comment:
Does this error message make sense? Is the return of find_connected_components not guaranteed?

Reply:
This function has some issues that require a re-write from scratch. That needs a separate PR that relies on this one.